### PR TITLE
RFC2: Initial integration of "GDAL SRS barn" work

### PR DESCRIPTION
**Pull request by rouault**
A HTML build of the RFC be found at http://even.rouault.free.fr/proj_cpp_api/rfc-2.html for easier reading

### CODE REVIEW

**Comment by hobu** [comment]
> It is not clear yet how this documentation can be integrated with the general PROJ documentation with the reStructuredText format.

PDAL uses a Sphinx extension called [Breathe](https://breathe.readthedocs.io/en/latest/) that allows Doxygen XML to be bound into Sphinx output. If that looks like too much effort, a simple static tree of Doxygen HTML would probably be sufficient to be pushed to GitHub pages.

**Comment by rouault** [comment]
@hobu Thanks for the hint ! I've managed a basic integration with Breathe. See http://even.rouault.free.fr/proj_cpp_api/rst_generated/html/development/reference/cpp/index.html

File: docs/source/community/rfc/rfc-2.rst
Comment (kbevers) [inline docs/source/community/rfc/rfc-2.rst]: An `i` to many

File: docs/source/community/rfc/rfc-2.rst
Comment (kbevers) [inline docs/source/community/rfc/rfc-2.rst]: > desired are of used

 Was this meant to say "desired area of use"?

File: docs/source/community/rfc/rfc-2.rst
Comment (kbevers) [inline docs/source/community/rfc/rfc-2.rst]: "desired"

File: docs/source/community/rfc/rfc-2.rst
Comment (kbevers) [inline docs/source/community/rfc/rfc-2.rst]: attribtes -> attributes

**Review by kbevers — COMMENTED** [review]
Even, this is an impressive write-up of what you've been doing over last few months. I definitely have a much better understanding of what you have done after reading this. 

As far as I can tell you haven't made any formal proposals in this RFC yet, is that correct? (It's a bit long - might have missed it :) ) I assume the end goal is to go through some serious reviewing of the code and subsequent merging into master?

**Comment by rouault** [comment]
> As far as I can tell you haven't made any formal proposals in this RFC yet, is that correct? (It's a bit long - might have missed it :) ) I assume the end goal is to go through some serious reviewing of the code and subsequent merging into master?

Ah, I thought that it was obvious that the implicit proposal was agreement for merging (I followed a bit the structure of similar GDAL or MapServer RFC). If people want to review, that would be great however that means reviewing nearly 30 000 lines of code (just speaking of stuff in include/proj and src/*.cpp)...

**Comment by kbevers** [comment]
Makes sense, of course. I think I was just tripped up by the concluding remark: "Not submitted to vote."

**Comment by rouault** [comment]
How should we proceed at that point ? Did people have enough time to make some form of review or are they just happy ? Call for a vote on the RFC ? (hum, some of it is already outdated by later changes, but the gist of it is still valid :-))

I'm not yet in an emergency to see this merged into master, but that's the ultimate goal, isn't it? I cannot exclude some changes (e.g adding a column in the DB structure) that would create significant diff changes, but things should hopefully stabilize at that point. One thing I'd note is that some of the new .cpp files tend to be quite large. I've no particular issue with that, and it can potentially enable compilers to generate better code (but that's arguable from my experience with the >1 week of optimizations I did), but if folks are really concerned about that, I could perhaps look at splitting a bit.

As far as merging, I think I'll do a single huge commit with all the work (keeping my original branch as it for historical purposes), to avoid polluting too much master history with a lot of intermediate steps

**Comment by hobu** [comment]
I'm 👍 on this and will defer to you if you think it is ready to merge. I'm excited to have it included.

**Comment by kbevers** [comment]
I'm with Howard here. If you feel it's time to merge I'm on board. Realistically it will take me anywhere from 6 to 12 months to get comfortable with the new parts of the code and we can't wait for that :) I think the overall structure looks sane, tests are plentyful and the docs are coming along nicely as well so now seems like a good time to merge. Squashing it all into one commit is probably the best way forward although it will be one mammoth of a commit!

Let's go for a vote on the RFC before merging. I think it will be good if you can touch up the sections that have become outdated.

File: docs/source/community/rfc/rfc-2.rst
Comment (schwehr) [inline docs/source/community/rfc/rfc-2.rst]: Strike `to be`.

File: docs/source/community/rfc/rfc-2.rst
Comment (schwehr) [inline docs/source/community/rfc/rfc-2.rst]: `dialect (we currently` -> `dialect.  PROJ currently ingests`

File: docs/source/community/rfc/rfc-2.rst
Comment (schwehr) [inline docs/source/community/rfc/rfc-2.rst]: Suggest adding `the` in here: `existing the code base`

**Review by schwehr — CHANGES_REQUESTED** [review]
How does this relate to pr #1040?  If there is a link (of course there is, correct?), be explicit about it somewhere within the text.

File: docs/source/community/rfc/rfc-2.rst
Comment (rouault) [reply docs/source/community/rfc/rfc-2.rst:34]: They are included in the Doxygen documentation of the repo, which is imported in the ReST doc, so they (the link to them) will be available in PROJ docu

File: docs/source/community/rfc/rfc-2.rst
Comment (rouault) [reply docs/source/community/rfc/rfc-2.rst]: Frenchism :-)

File: docs/source/community/rfc/rfc-2.rst
Comment (rouault) [reply docs/source/community/rfc/rfc-2.rst:46]: you can't. This is not yet an OGC official standard, so you can't have access to it if not a OGC member. Hopefully this will be released in a few weeks / months. But looking at the existing ISO-19111:2007 ( https://portal.opengeospatial.org/files/?artifact_id=39049 ) will give you a quite good picture already

File: docs/source/community/rfc/rfc-2.rst
Comment (rouault) [reply docs/source/community/rfc/rfc-2.rst:50]: no. This is also an OGC public standard. See above link

File: docs/source/community/rfc/rfc-2.rst
Comment (rouault) [reply docs/source/community/rfc/rfc-2.rst]: You're nitpicking. Frenchism

File: docs/source/community/rfc/rfc-2.rst
Comment (rouault) [reply docs/source/community/rfc/rfc-2.rst:151]: It is more convenient to have it close to the other includes, as it is needed for the public C++ API

File: docs/source/community/rfc/rfc-2.rst
Comment (kbevers) [reply docs/source/community/rfc/rfc-2.rst]: [Conversions](https://proj4.org/operations/conversions/index.html) and [transformations](https://proj4.org/operations/transformations/index.html) are defined in the existing docs. I don't think it is necessary to repeat those, but links would be helpful.

File: docs/source/community/rfc/rfc-2.rst
Comment (rouault) [reply docs/source/community/rfc/rfc-2.rst:152]: This is a discussion I started a few months ago in http://lists.maptools.org/pipermail/proj/2018-May/008248.html.  Basically shared_ptr are easier to deal with and make it error prone for API users (even internally)

File: docs/source/community/rfc/rfc-2.rst
Comment (schwehr) [reply docs/source/community/rfc/rfc-2.rst]: Please fix.   I'm sure most people wouldn't notice, but (personally) it's not a nit w.r.t my ability to read.  My brain is looking for extra meaning that isn't there.

File: docs/source/community/rfc/rfc-2.rst
Comment (rouault) [reply docs/source/community/rfc/rfc-2.rst]: ISO 19111 terminology. Short introduction at https://proj4.org/operations/index.html

File: docs/source/community/rfc/rfc-2.rst
Comment (rouault) [reply docs/source/community/rfc/rfc-2.rst]: in unspecified geodetic CRS, so long lat

File: docs/source/community/rfc/rfc-2.rst
Comment (rouault) [reply docs/source/community/rfc/rfc-2.rst]: Not sure to understand your question. I'm talking about https://proj4.org/resource_files.html#transformation-grids if that's not clear

File: docs/source/community/rfc/rfc-2.rst
Comment (schwehr) [reply docs/source/community/rfc/rfc-2.rst]: Then please change.  It hurts readability for others

File: docs/source/community/rfc/rfc-2.rst
Comment (schwehr) [reply docs/source/community/rfc/rfc-2.rst:34]: Cool.  Is there a way to link to those?

File: docs/source/community/rfc/rfc-2.rst
Comment (rouault) [reply docs/source/community/rfc/rfc-2.rst]: RST formatting to separate the command, and the summarized output (returns 78 results) I did to avoid putting tens of pages of results

File: docs/source/community/rfc/rfc-2.rst
Comment (schwehr) [reply docs/source/community/rfc/rfc-2.rst]: It's sorta clear to me, but me of 6 years ago would have no idea what you are referring to.

File: docs/source/community/rfc/rfc-2.rst
Comment (schwehr) [reply docs/source/community/rfc/rfc-2.rst]: I definitely got what you are trying to do (and I 100% agree), but can you restate it something like:

    The following command currently returns 78 results:

        projinfo -s EPSG:4267 -t EPSG:4326 --summary --spatial-test intersects

File: docs/source/community/rfc/rfc-2.rst
Comment (rouault) [reply docs/source/community/rfc/rfc-2.rst]: Yeah, I'm trying to rephrase the Japanese example

File: docs/source/community/rfc/rfc-2.rst
Comment (rouault) [reply docs/source/community/rfc/rfc-2.rst:400]: This is not actually true. authority_list is a view, but perhaps I should rename it as _view

File: docs/source/community/rfc/rfc-2.rst
Comment (rouault) [reply docs/source/community/rfc/rfc-2.rst:403]: This was to make it a bit clearer what the generic metadata term means here

File: docs/source/community/rfc/rfc-2.rst
Comment (rouault) [reply docs/source/community/rfc/rfc-2.rst:650]: What is the worry exactly ?

File: docs/source/community/rfc/rfc-2.rst
Comment (schwehr) [reply docs/source/community/rfc/rfc-2.rst:400]: I wasn't thinking about the names of the table or view (just thinking of the text that follows), but now you have me wondering what's better.  I don't see a reason to name it `_view`.  Doesn't really matter, yes?  Be it table or view, I believe it works the same from a query perspective.  Or go all out and have fun: `authority_list_table_view_builder_generator_non_authorative_catalog_oh_my_brain`

File: docs/source/community/rfc/rfc-2.rst
Comment (schwehr) [reply docs/source/community/rfc/rfc-2.rst:650]: What if I have another part of my tree that already uses nn?  What if it uses a different version of nn.hpp?  What if a distributions policy is to pull out and separately package this sort of thing?  The classic ODR https://en.wikipedia.org/wiki/One_Definition_Rule

Comment (rouault) [reply docs/source/community/rfc/rfc-2.rst:650]: This is a header only package that will be in ${prefix}/include/proj, so you can't include it by accident if you just do -I{prefix}/include (you'd need to do #include <proj/nn.hpp>). As far as ODR, this is a header-only dependency that defines template classes that take as template argument PROJ classes, so even if you use a different version of it in your program, the actual instanciation of classes will not be the same.

**Comment by rouault** [comment]
@schwehr Does https://github.com/OSGeo/proj.4/pull/1149/commits/ed0ee0bc9f0bada66eb42f6d0ed602d12558cf7c and my inline answers capture well your review?

File: docs/source/community/rfc/rfc-2.rst
Comment (rouault) [reply docs/source/community/rfc/rfc-2.rst:400]: The initial reason for _view was that I had a "coordinate_operation" table (that just listed the auth_name, code, table_name) and that served as a foreign key target for the "helmert_transformation", "grid_transformation", etc. table so that there is no duplication of (auth_name, code) between those "derived" tables. The coordinate_operation_view was actually a view selecting in those "derived" tables and having much more information.
I've finally dropped the coordinate_operation table since that was a bit painful and not user friendly, and replaced it with trigger checks.
I'm not yet sure if the _view naming convention is better or not. It has the advantage of indicating you cannot insert into it.

**Comment by kbevers** [comment]
Please cherry-pick this commit into 5.2 so the website get updated.

**Comment by rouault** [comment]
> Please cherry-pick this commit into 5.2 so the website get updated.

done

**Pull request merged**
